jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java

```diff
@@ -1052,18 +1058,42 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
                     case COOKIE:
                     case CACHE_CONTROL:
                     case USER_AGENT:
-                        addToConnectionTrie = _fieldCache != null && _field == null;
+                        addToFieldCache = _field == null;
                         break;
 
                     default:
                         break;
                 }
 
-                if (addToConnectionTrie && !_fieldCache.isFull() && _header != null && _valueString != null)
+                // Cache field?
+                if (addToFieldCache && _header != null && _valueString != null)
                 {
-                    if (_field == null)
-                        _field = new HttpField(_header, caseInsensitiveHeader(_headerString, _header.asString()), _valueString);
-                    _fieldCache.put(_field);
+                    if (_fieldCache != null)
+                    {
+                        // If a non full cache already exists, then we can add to it.
+                        if (!_fieldCache.isFull())
+                        {
+                            if (_field == null)
+                                _field = new HttpField(_header, caseInsensitiveHeader(_headerString, _header.asString()), _valueString);
+                            _fieldCache.put(_field);
+                        }
+                    }
+                    else if (_cacheableFields != NO_CACHE)
+                    {
+                        if (_handler.getHeaderCacheSize() <= 0 || (_version != null && _version.getVersion() != HttpVersion.HTTP_1_1.getVersion()))
+                            // Don't cache any fields
+                            _cacheableFields = NO_CACHE;
+                        else
+                        {
+                            // This must be the first request seen by this parser, so just create a simple list of
+                            // headers that can later to converted to _fieldCache if another request comes.
+                            if (_cacheableFields == null)
+                                _cacheableFields = new ArrayList<>();
+                            if (_field == null)
+                                _field = new HttpField(_header, caseInsensitiveHeader(_headerString, _header.asString()), _valueString);
+                            _cacheableFields.add(_field);
```

Comment:
In this branch we have a cache size > 0, we are HTTP/1.1 and there is no `Connection: close`, so chances are that this parser will see another request, likely more than 1.
I would remove entirely the `_cacheableFields` handling, and just use the trie -- we have already optimized its creation.

Reply:
The problem is that we don't know if we will get a `Connection: close` header before or after any cacheable fields.... which is why I want the list.
If Connection:close is the last field, we could have allocated a big cache, done the work to fill it up and then find out at the last moment that it is not persistent.

Reply:
A client sending HTTP/1.1 first requests with `Connection: close` is not a common case, so you are right that we may see the `Connection: close` header after we have created the trie, but that would be a rare case for which I don't think the extra complexity is warranted.

Reply:
I don't know why you think it is a rare case?   For clients sending a 1 time HTTP/1.1 request we have no data on if the cacheable fields like`User-Agent` and `Cookie` will come before or after a `Connection: close`.

Creating the list means that we have minimal cost for all 1 request connections and only slight extra cost for the first request of a multiple request connections.
Not creating the list means that we have significant cost for an unknown proportion of 1 request connections and slight saving on multiple request connections.

So while I don't like the list that much, I think it is best average solution.

An alternative would be to always make the list, then when a new request comes in we can look at what we have cached and what is in the list.  We can then really see if fields are likely to be the same or if they are actually changing.  We could then make smarter decisions about full caches and this perhaps have smaller caches.   I have to think about this a bit more, but having the list on the first request is a step towards this if it is workable.

Reply:
If we always created the list of cacheable fields and then only created/updated the actual cache on the next request then we would have a bit more info.
If the cache is perfect (ie contains all the cacheable fields that exist and they never change between requests), then on subsequent requests the cacheableField list will be empty.
On the 2nd and later requests if there are fields in the list then there are several possibilities
 + If the field name is not in the cache, then either the cache is full or it is a new field.   In the later case we can add the field.
 + If the field name is in the cache, then it is a changed value for that field.  Either that field is going to continually change (eg digest auth), it has changed once and will be the same afterwards, it is one of several values that may alternate (eg User-Agent on an aggregated connection).   

Hmmm not sure there is anything smart we can do with this information.  I think it is just add the filed to the cache if there is space.   So always having the list might save a little bit of work on the Nth request of a N request stream.  Not worth it.

So I still think the list of the 1st request a reasonable compromise.